src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/Messager.java

```diff
@@ -383,21 +399,238 @@ public void printErrorWarningCounts() {
         }
     }
 
-    private void report(DiagnosticType type, String pos, String msg) {
-        switch (type) {
-            case ERROR:
-            case WARNING:
-                Object prefix = (pos == null) ? programName : pos;
-                report(javadocDiags.create(type, null, null, "msg", prefix, msg));
-                break;
+    /**
+     * Reports a diagnostic message.
+     *
+     * @param kind    the kind of diagnostic
+     * @param ds      the diagnostic source
+     * @param dp      the diagnostic position
+     * @param message the message
+     */
+    private void report(Diagnostic.Kind kind, DiagnosticSource ds, DiagnosticPosition dp, String message) {
+        report(getDiagnosticType(kind), getDiagnosticFlags(kind), ds, dp, message);
+    }
+
+    /**
+     * Reports a diagnostic message.
+     *
+     * @param dt      the diagnostic type
+     * @param ds      the diagnostic source
+     * @param dp      the diagnostic position
+     * @param message the message
+     */
+    private void report(DiagnosticType dt, DiagnosticSource ds, DiagnosticPosition dp, String message) {
+        report(dt, EnumSet.noneOf(DiagnosticFlag.class), ds, dp, message);
+    }
+
+    /**
+     * Reports a diagnostic message, with diagnostic flags.
+     * For javadoc, the only flag that is used is {@code MANDATORY_WARNING}, and only
+     * because in principle the public API supports it via {@code Kind.MANDATORY_WARNING}.
+     * javadoc itself does generate mandatory warnings.
+     *
+     * This is the primary low-level wrapper around the underlying {@code Log.report}.
+     * Because we already have a localized message, we use wrapper resources (just {@code {0}})
+     * to wrap the string. The current behavior is one wrapper per diagnostic type.
+     * We could improve this by subtyping {@code DiagnosticInfo} to modify the resource key used.
+     *
+     * {@code Log} reports all diagnostics to the corresponding writer, which defaults
+     * to the "error" stream, when using the two-stream constructor. That doesn't work
+     * for javadoc, which has historically written notes to the "output" stream, because
+     * the public API used by doclets does not provide for more detailed control.
+     * Therefore, for now, javadoc continues to use the (deprecated) three-stream
+     * constructor, with the {@code NOTE} stream set to the "output" stream.
+     *
+     * {@code Log} reports all notes with a "Note:" prefix. That's not good for the
+     * standard doclet, which uses notes to report the various "progress" messages,
+     * such as  "Generating class ...".  Therefore, for now, we detect and report those
+     * messages directly. (A better solution would be to expose access to the output
+     * and error streams via {@code DocletEnvironment}.
```

Comment:
Sentence: either remove "(" or add ")".

Reply:
Add ")" and change DocletEnvironment to Reporter for now, and to delete the sentence when we add streams to Reporter.